db/compaction/compaction_outputs.cc

```diff
@@ -571,39 +596,167 @@ Status CompactionOutputs::AddRangeDels(
+    //   output file always contains a point key. Each new compaction output
+    //   file is created if there is a point key for which ShouldStopBefore()
+    //   returns true, and the point key would be added to the new compaction
+    //   output file. So each new compaction file always contains a point key.
+    //   So Case 1 only happens when CompactionIterator does not emit any
+    //   point key.
+    //
+    //   To show (1) meta.smallest <= meta.largest:
+    //   Since the compaction output is range tombstone only, `lower_bound` and
+    //   `upper_bound` are either null or comp_start/end_user_key respectively.
+    //   According to how UpdateBoundariesForRange() is implemented, it blindly
+    //   updates meta.smallest and meta.largest to smallest_candidate and
+    //   largest_candidate the first time it is called. Subsequently, it
+    //   compares input parameter with meta.smallest and meta.largest. So we
+    //   only need to show smallest_candidate <= largest_candidate the first
+    //   time UpdateBoundariesForRange() is called. We assume
+    //   comp_start_user_key < comp_end_user_key, if provided. We assume that
+    //   tombstone_start < tombstone_end. This assumption is based on that each
+    //   fragment in FragmentedTombstoneList has start_key < end_key (user_key)
+    //   and that FragmentedTombstoneIterator::Tombstone() returns the pair
+    //   (start_key@tombstone_seqno with op_type kTypeRangeDeletion, end_key).
+    //   The logic in this loop sets smallest_candidate to
+    //   max(tombstone_start.user_key, comp_start_user_key)@tombstone.seq_ with
+    //   op_type kTypeRangeDeletion, largest_candidate to
+    //   min(tombstone_end.user_key, comp_end_user_key)@kMaxSequenceNumber with
+    //   op_type kTypeRangeDeletion. When a bound is null, there is no
+    //   truncation on that end. To show that smallest_candidate <=
+    //   largest_candidate, it suffices to show tombstone_start.user_key <
+    //   comp_end_user_key (if not null) AND comp_start_user_key (if not null) <
+    //   tombstone_end.user_key. Since the file has no point key,
+    //   `has_overlapping_endpoints` is false. In the first sanity check of this
+    //   for-loop, we compare tombstone_start.user_key against upper_bound =
+    //   comp_end_user_key, and only proceed if tombstone_start.user_key <
+    //   comp_end_user_key. We assume FragmentedTombstoneIterator::Seek(k) lands
+    //   on a tombstone with end_key > k. So the call it->Seek(*lower_bound)
+    //   above implies compact_start_user_key < tombstone_end.user_key.
+    //
+    //   To show (2) [tombstone_start, tombstone_end] and [meta.smallest,
+    //   meta.largest] overlaps (after the call to UpdateBoundariesForRange()):
+    //   We show that [smallest_candidate, largest_candidate] overlaps with
+    //   [meta.smallest, meta.largest], and since tombstone_start <=
+    //   smallest_candidate <= largest_candidate <= tombstone_end, (2) holds
```

Comment:
I'm a bit lost on smallest_candidate <= largest_candidate. What's the source of this guarantee? Is it actually stronger, like smallest_candidate.user_key < largest_candidate.user_key?

Reply:
Sorry it's buried in proof for (1) where we show smallest_candidate <= largest_candidate and it implies meta.smallest <= meta.largest. I'll make it clearer.

> Is it actually stronger, like smallest_candidate.user_key < largest_candidate.user_key?

Yes, in this case. This does not hold for Case 2 though, since we can have smallest_candidate > largest_candidate in that case.